### PR TITLE
import and export transcripts

### PR DESCRIPTION
**Pull request by muhammad-ammar**
[EDUCATOR-1216](https://openedx.atlassian.net/browse/EDUCATOR-1216)

### CODE REVIEW

**Comment by muhammad-ammar** [comment]
@muzaffaryousaf @Qubad786 This is ready for review.

File: edxval/api.py [hunk 1]
```diff
@@ -652,22 +652,36 @@ def copy_course_videos(source_course_id, destination_course_id):
...
         course_id (str): The ID of the course with which this video is associated
+        external (bool): True if edx_video_id is not present else False
 
     Returns:
         An lxml video_asset element containing export data
 
     Raises:
         ValVideoNotFoundError: if the video does not exist
     """
+    if not isinstance(video_ids, list):
```
Comment (muzaffaryousaf) [inline edxval/api.py]: can we also check that video_ids is not empty.

File: edxval/api.py [hunk 2]
```diff
@@ -652,22 +652,36 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+        raise ValueError('"video_ids" argument must be a list')
```
[the remark this answers is not in the document]
Comment (muzaffaryousaf) [reply edxval/api.py]: '"video_ids" argument must be a non-empty list'

File: edxval/api.py [hunk 3]
```diff
@@ -652,22 +652,36 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+
+    # As val does not store external videos, so construct transcripts information only.
+    if external:
+        video_el = Element('video_asset')
+        return create_transcripts_xml(video_ids, video_el)
+    else:
```
Comment (muzaffaryousaf) [inline edxval/api.py]: We don't need else part.

File: edxval/api.py [hunk 4]
```diff
@@ -652,22 +652,36 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+        # for an internal video, first video id must be edx_video_id
+        video_id = video_ids[0]
```
Comment (muzaffaryousaf) [inline edxval/api.py]: Move this out of else.

File: edxval/tests/test_api.py [hunk 5]
```diff
@@ -1020,18 +1082,44 @@ def assert_invalid_import(self, xml, course_id=None):
...
 
+    def assert_transcripts(self, video_id, expected_transcripts):
+        """
+        Compare `received` with `expected` and assert if not equal
+        """
+        # Verify total number of expected transcripts for a video
+        video_transcripts = VideoTranscript.objects.filter(video_id=video_id)
+        self.assertEqual(video_transcripts.count(), len(expected_transcripts))
+
+        # Verify data fro each transcript
```
Comment (muzaffaryousaf) [inline edxval/tests/test_api.py]: fro --> for

File: edxval/api.py [hunk 6]
```diff
@@ -652,22 +652,36 @@ def copy_course_videos(source_course_id, destination_course_id):
...
 
-def export_to_xml(edx_video_id, course_id=None):
+def export_to_xml(video_ids, course_id=None, external=False):
     """
-    Exports data about the given edx_video_id into the given xml object.
+    Exports data for a video into an xml object.
 
-    Args:
-        edx_video_id (str): The ID of the video to export
+    NOTE: For external video ids, only transcripts information will be added into xml.
```
Comment (Qubad786) [inline edxval/api.py]: Additional note: If external=True, then `edx_video_id` is going to be on first index of the list.

File: edxval/tests/test_api.py [hunk 7]
```diff
@@ -905,6 +905,14 @@ def setUp(self):
             **constants.ENCODED_VIDEO_DICT_HLS
         )
 
+        VideoTranscript.objects.create(**constants.VIDEO_TRANSCRIPT_CUSTOM)
+        video_transcript = dict(constants.VIDEO_TRANSCRIPT_CUSTOM, **{'language_code': 'ar'})
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: `dict(constants.VIDEO_TRANSCRIPT_CUSTOM, language_code=u'ar'})`

File: edxval/tests/test_api.py [hunk 8]
```diff
@@ -1020,18 +1082,44 @@ def assert_invalid_import(self, xml, course_id=None):
...
+            language_code = expected_transcript['language_code']
+            expected_transcript['name'] = expected_transcript.pop('transcript')
+
+            # get the imported transcript and rename `url` key
+            received = api.TranscriptSerializer(
+                VideoTranscript.objects.get(video_id=video_id, language_code=language_code)
+            ).data
+            received['name'] = received.pop('url')
+
+            self.assertEqual(received, expected_transcript)
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: can we use `self.assertDictEqual`?

**Comment by muhammad-ammar** [comment]
@muzaffaryousaf @Qubad786 Feedback addressed.

File: edxval/api.py [hunk 9]
```diff
@@ -765,3 +819,24 @@ def import_from_xml(xml, edx_video_id, course_id=None):
...
+def create_transcript_objects(xml):
+    """
+    Create VideoTranscript objects.
+
+    Arguments:
+        xml (Element): lxml Element object
+    """
+    for transcript in xml.findall('.//transcripts/transcript'):
+        try:
+            VideoTranscript.create_or_update(
```
Comment (Qubad786) [inline edxval/api.py]: I think we decided to not to overwrite/update existing video transcripts?

File: edxval/tests/test_api.py [hunk 10]
```diff
@@ -1157,6 +1274,69 @@ def test_invalid_course_id(self):
...
+        video_id = 'little-star'
+        transcript_xml = '<transcript file_name="wow.srt" language_code="en" file_format="srt" provider="Cielo24"/>'
+        xml = etree.fromstring("""
+            <video_asset>
+                <transcripts>
+                    {transcript_xml}
+                    <transcript file_name="wow.sjson" language_code="de" file_format="sjson" provider='3PlayMedia' video_id="{video_id}"/>
+                </transcripts>
+            </video_asset>
+        """.format(transcript_xml=transcript_xml, video_id=video_id))
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: we could just put the following directly instead for formatting it:
`'<transcript file_name="wow.srt" language_code="en" file_format="srt" provider="Cielo24"/>'`

Comment (muhammad-ammar) [reply edxval/tests/test_api.py]: `transcript_xml` is needed at other places.

File: edxval/api.py (hunk 9 above)
Comment (muhammad-ammar) [reply edxval/api.py]: Discussed with @muzaffaryousaf and decided that we will update existing transcripts upon import.

Do you see any issues?

File: edxval/api.py [hunk 11]
```diff
@@ -7,6 +7,7 @@
 from enum import Enum
 
 from django.core.exceptions import ObjectDoesNotExist, ValidationError
+from lxml import etree
 from lxml.etree import Element, SubElement
```
[the remark this answers is not in the document]
Comment (muhammad-ammar) [reply edxval/api.py]: I don't think we can merge them.

[the remark this answers is not in the document]
Comment (Qubad786) [reply edxval/api.py]: I meant keeping on of them :-p

File: edxval/tests/test_api.py (hunk 10 above)
Comment (Qubad786) [reply edxval/tests/test_api.py]: Ah right!

File: edxval/api.py (hunk 9 above)
Comment (Qubad786) [reply edxval/api.py]: users will be able to manipulate existing transcripts by editing the xml.

File: edxval/tests/test_api.py [hunk 12]
```diff
@@ -1214,6 +1223,10 @@ def test_existing_video(self, course_id):
         course_video = video.courses.get(course_id=course_id)
         self.assertTrue(course_video.video_image.image.name, self.image_name)
 
+        self.assert_transcripts(
+            constants.VIDEO_DICT_FISH["edx_video_id"],
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: nit: use single quotes

File: edxval/tests/test_api.py [hunk 13]
```diff
@@ -1197,8 +1200,14 @@ def test_existing_video(self, course_id):
                     "profile": "mobile",
                 },
             ],
-            image=self.image_name
+            image=self.image_name,
+            video_transcripts=[transcript_data]
         )
+
+        # there must not be any transcript before import
+        with self.assertRaises(VideoTranscript.DoesNotExist):
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: nit: use single quotes

File: edxval/tests/test_api.py [hunk 14]
```diff
@@ -1197,8 +1200,14 @@ def test_existing_video(self, course_id):
...
+            VideoTranscript.objects.get(video_id=constants.VIDEO_DICT_FISH["edx_video_id"])
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: nit: use single quotes

**Pull request merged**